crates/agent-sql/src/directives/beta_onboard.rs

```diff
@@ -41,39 +41,86 @@ pub async fn tenant_exists(
+) -> sqlx::Result<ProvisionedTenant> {
     let prefix = format!("{tenant}/");
 
-    sqlx::query!(
-        r#"with s1 as (
+    let provisioned = sqlx::query_as!(
+        ProvisionedTenant,
+        r#"with
+        accounts_root_user as (
+            -- Precondition: the accounts root user must exist.
+            -- Use a sub-select to select either one match or an explicit null row,
+            -- which will then fail a not-null constraint.
+            select (select id from auth.users where email = $4 limit 1) as accounts_id
+        ),
+        grant_user_admin_to_tenant as (
             insert into user_grants (user_id, object_role, capability, detail) values
                 ($1, $2, 'admin', $3)
         ),
-        s2 as (
+        grant_to_tenant as (
             insert into role_grants (subject_role, object_role, capability, detail) values
-                ($2, 'ops/' || $2, 'read', $3)
+                ($2, $2, 'write', $3),              -- Tenant specs may write to other tenant specs.
+                ($2, 'ops/' || $2, 'read', $3),     -- Tenant may read `ops/$tenant/...` collections.
+                ($2, 'estuary/public/', 'read', $3) -- Tenant may read `estuary/pubic/` collections.
+        ),
+        create_storage_mappings as (
+            insert into storage_mappings (catalog_prefix, spec, detail) values
+                ($2, '{"stores": [{"provider": "GCS", "bucket": "estuary-trial"}]}', $3),
+                ('recovery/' || $2, '{"stores": [{"provider": "GCS", "bucket": "estuary-trial"}]}', $3)
+        ),
+        -- Create a draft for provisioned catalog specifications owned by the accounts root user.
+        -- It will be filled out later but within this same transaction.
+        -- Then queue a publication of that draft, also owned by the accounts root user.
+        create_draft as (
+            insert into drafts (user_id)
+                select accounts_id from accounts_root_user
+                returning drafts.id as draft_id
+        ),
+        create_publication as (
+            insert into publications (user_id, draft_id)
+                select accounts_id, draft_id from create_draft, accounts_root_user
         )
-        insert into storage_mappings (catalog_prefix, spec, detail) values
-            ($2, '{"stores": [{"provider": "GCS", "bucket": "estuary-trial"}]}', $3),
-            ('recovery/' || $2, '{"stores": [{"provider": "GCS", "bucket": "estuary-trial"}]}', $3)
+        select draft_id as "draft_id: Id" from create_draft;
         "#,
-        user_id as Uuid,
-        prefix as String,
+        tenant_user_id as Uuid,
+        &prefix as &str,
         detail.clone() as Option<String>,
+        accounts_user_email as &str,
     )
-    .execute(txn)
+    .fetch_one(&mut *txn)
+    .await?;
+
+    // Create partition of task_stats which will home all stats of the tenant.
+    sqlx::query(&format!(
+        r#"
+        create table task_stat_partitions."{tenant}_stats"
```

Comment:
It wasn't as much of a concern above because of bound input parameters, but since we're directly templating variable into SQL here, do we have to worry about injection? Can I ask for a `tenant` of `hello"; drop table users; --`, or some such?

Reply:
No, at this stage `tenant` has had a lot of validation. It's unicode letters, decimals, `_`, `-`, `.` and that's it.